File: spatial-lib/src/cmr/spatial/geodetic_ring.clj
(ns cmr.spatial.geodetic-ring
  (:require
   [cmr.common.dev.record-pretty-printer :as record-pretty-printer]
   [cmr.common.util :as util]
   [cmr.spatial.arc :as a]
   [cmr.spatial.conversion :as c]
   [cmr.spatial.derived :as d]
   [cmr.spatial.math :refer :all]
   [cmr.spatial.mbr :as mbr]
   [cmr.spatial.point :as p]
   [primitive-math])
  (:import cmr.spatial.arc.Arc))

(primitive-math/use-primitive-operators)

(def ^:const ^:private EXTERNAL_POINT_PRECISION
  "Defines the precision in degrees to use when generating external points to a ring and when
  comparing a point for use with an external point."
  4)

(defrecord GeodeticRing
  [;; The points that make up the ring. Points must be in counterclockwise order. The last point
   ;; must match the first point.
   points
   ;;
   ;; Derived fields
   ;;
   ;; A set of the unique points in the ring.
   ;; This should be used as opposed to creating a set from the points many times over which is expensive.
   point-set
   ;; The arcs of the ring
   arcs
   ;; This attribute contains the rotation direction for the ring. Rotation direction will be one of
   ;; :clockwise, :counter-clockwise, or :none to indicate the point order direction.
   ;; * :clockwise indicates the points are listed in a clockwise order around a center point.
   ;; * :counter-clockwise indicates the points are listed in a counter clockwise order around a center point.
   ;; * :none indicates the point order is around the earth like a belt.
   ;; Depending on the order it could contain the south or north pole.
   course-rotation-direction
   ;; true if ring contains north pole
   contains-north-pole
   ;; true if ring contains south pole
   contains-south-pole
   ;; the minimum bounding rectangle
   mbr
   ;; Three points that are not within the ring. These are used to test if a point is inside or
   ;; outside a ring. We generate multiple external points so that we have a backup if one external
   ;; point is antipodal to a point we're checking is inside a ring.
   external-points])

(record-pretty-printer/enable-record-pretty-printing GeodeticRing)

(defn- arcs-and-arc-intersections
  "Returns the set of intersection points between the arc and the list of arcs. "
  [arcs other-arc]
  (persistent!
    (reduce (fn [s arc]
              (if-let [[point1 point2] (seq (a/intersections arc other-arc))]
                ;; Round the points. If the crossing arc passes through a point on the ring the
                ;; intersection algorithm will result in two very, very close points. By rounding to
                ;; within an acceptable range they'll be seen as the same point.
                (let [s (conj! s (p/round-point p/INTERSECTION_POINT_PRECISION point1))]
                  (if point2
                    (conj! s (p/round-point p/INTERSECTION_POINT_PRECISION point2))
                    s))
                s))
            (transient #{})
            arcs)))

(defn- choose-external-point
  "Finds an external point to use with the ring when testing for coverage of the given point. An
  external point cannot be the same as the point or antipodal to the given point."
  [ring point]
  (let [[ep1 ep2 ep3] (:external-points ring)
        point (p/round-point EXTERNAL_POINT_PRECISION point)
        antipodal-point (p/antipodal point)]
    (cond
      (and (not= ep1 point) (not= ep1 antipodal-point)) ep1
      (and (not= ep2 point) (not= ep2 antipodal-point)) ep2
      (and (not= ep3 point) (not= ep3 antipodal-point)) ep3
      :else
      (throw (Exception. (str "Could not find external point to use to check if ring covers"
                              " point. Ring: " (pr-str ring) " point: " (pr-str point)))))))

(defn covers-point?
  "Determines if a ring covers the given point. The algorithm works by counting the number of times
  an arc between the point and a known external point crosses the ring. An even count means the point
  is external. An odd count means the point is inside the ring."
  [ring point]
  ;; The pre check is necessary for rings which might contain both north and south poles
  {:pre [(> (count (:external-points ring)) 0)]}

  (or (and (:contains-north-pole ring) (p/is-north-pole? point))
      (and (:contains-south-pole ring) (p/is-south-pole? point))
      ;; Only do real intersection if the mbr covers the point.
      (when (mbr/geodetic-covers-point? (:mbr ring) point)
        (if ((:point-set ring) point)
          true ; The point is actually one of the rings points
          ;; otherwise we'll do the real intersection algorithm
          (let [external-point (choose-external-point ring point)
                ;; Create the test arc
                crossing-arc (a/arc point external-point)
                intersections (arcs-and-arc-intersections (:arcs ring) crossing-arc)]
            (or (odd-long? (count intersections))
                ;; if the point itself is one of the intersections then the ring covers it
                (intersections point)))))))

(defn- arcs->course-rotation-direction
  "Calculates the rotation direction of the arcs of a ring. Will be one of :clockwise,
  :counter-clockwise, or :none.

  It works by calculating the number of degrees of turning that the ring does. It gets the initial
  and ending course from each arc. It determines how many degrees each turn is. Turns to the left,
  counter clockwise, are positive. Turns to the right, clockwise, are negative. This adds all of the
  differences together to get the net bearing change while traveling around the ring. A normal
  counter clockwise ring will be approximately 360 degrees of turn. A clockwise ring will be -360.
  A ring around a pole will be approximately 0 net degrees turn. If a ring crosses or has a point on
  a single pole then the sum will be -180 or 180. If a ring crosses both poles then the sum will be
  0."
  [arcs]
  ;; Gets a list of the arc initial and ending courses to show all the angles that are travelled on
  ;; throughout the ring.
  (let [courses (loop [courses (transient []) arcs arcs]
                  (if (empty? arcs)
                    (persistent! courses)
                    (let [^Arc a (first arcs)]
                      (recur (-> courses
                                 (conj! (.initial_course a))
                                 (conj! (.ending_course a)))
                             (rest arcs)))))
        ;; Add the first turn angle on again to complete the turn
        courses (conj courses (first courses))]
    (rotation-direction courses)))

(defn ring
  "Creates a new ring with the given points. If the other fields of a ring are needed. The
  calculate-derived function should be used to populate it."
  [points]
  (->GeodeticRing (mapv p/with-geodetic-equality points) nil nil nil nil nil nil nil))

(defn contains-both-poles?
  "Returns true if a ring contains both the north pole and the south pole"
  [ring]
  (and (:contains-north-pole ring)
       (:contains-south-pole ring)))

(defn point-order?
  "Returns true if the points are in counter clockwise order"
  [ring]
  (= (:course-rotation-direction ring) :clockwise))

(defn ring->arcs
  "Determines the arcs from the points in the ring."
  [^GeodeticRing ring]
  (or (.arcs ring)
      (a/points->arcs (.points ring))))

(defn ring->pole-containment
  "Returns the ring with north and south pole containment determined"
  [^GeodeticRing ring]
  (if (:course-rotation-direction ring)
    ring
    (let [arcs (ring->arcs ring)
          points (.points ring)
          course-rotation-direction (arcs->course-rotation-direction arcs)
          ;; The net rotation direction of the longitudes of the ring around the earth if looking
          ;; down on the north pole
          lon-rotation-direction (->> points (mapv :lon) rotation-direction)

          contains-north-pole (or (some p/is-north-pole? points)
                                  (some a/crosses-north-pole? arcs)
                                  (and (= :none course-rotation-direction)
                                       (= :counter-clockwise lon-rotation-direction)))

          contains-south-pole (or (some p/is-south-pole? points)
                                  (some a/crosses-south-pole? arcs)
                                  (and (= :none course-rotation-direction)
                                       (= :clockwise lon-rotation-direction)))]
      (assoc ring
             :course-rotation-direction course-rotation-direction
             :contains-north-pole contains-north-pole
             :contains-south-pole contains-south-pole))))

(defn ring->point-order
  "Returns the direction of the arcs of a ring, either :clockwise, :counter-clockwise, or :none."
  [^GeodeticRing ring]
  (arcs->course-rotation-direction (ring->arcs ring)))

(defn ring->mbr
  "Determines the mbr from the points in the ring."
  [^GeodeticRing ring]
  (or (.mbr ring)
      (let [arcs (ring->arcs ring)
            {:keys [contains-north-pole contains-south-pole]} (ring->pole-containment ring)
            br (reduce (fn [br ^Arc arc]
                         (let [mbr1 (.mbr1 arc)
                               br (if br
                                    (mbr/union br mbr1)
                                    mbr1)
                               mbr2 (.mbr2 arc)]
                           (if mbr2
                             (mbr/union br mbr2)
                             br)))
                       nil
                       arcs)
            br (if (and contains-north-pole
                        (not-any? p/is-north-pole? (:points ring))
                        (not-any? a/crosses-north-pole? arcs))
                 (mbr/mbr -180.0 90.0 180.0 (:south br))
                 br)]
        (if (and contains-south-pole
                 (not-any? p/is-south-pole? (:points ring))
                 (not-any? a/crosses-south-pole? arcs))
          (mbr/mbr -180.0 (:north br) 180.0 -90.0)
          br))))


(defn ring->external-points
  "Determines external points that are not in the ring."
  [^GeodeticRing ring]
  (let [br (ring->mbr ring)
        {:keys [contains-north-pole contains-south-pole]} (ring->pole-containment ring)]
    (if (and contains-north-pole contains-south-pole)
      ;; Cannot determine external points of a ring which contains both north and south poles
      ;; This is an additional feature which could be added at a later time.
      []
      (mapv #(p/round-point EXTERNAL_POINT_PRECISION %) (mbr/external-points br)))))

(extend-protocol d/DerivedCalculator
  cmr.spatial.geodetic_ring.GeodeticRing
  (calculate-derived
    [^GeodeticRing ring]
    (if (.arcs ring)
      ring

      (as-> ring ring
            (assoc ring :point-set (set (:points ring)))
            (assoc ring :arcs (ring->arcs ring))
            (ring->pole-containment ring)
            (assoc ring :mbr (ring->mbr ring))
            (assoc ring :external-points (ring->external-points ring))))))

File: spatial-lib/src/java/cmr/spatial/internal/ring/GeodeticRing.java
package cmr.spatial.internal.ring;

import cmr.spatial.shape.Point;
import cmr.spatial.shape.Mbr;
import cmr.spatial.internal.arc.Arc;
import cmr.spatial.internal.arc.ArcLineSegmentIntersections;
import cmr.spatial.math.MathUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Represents a geodetic (spherical) ring - a closed sequence of points on a sphere
 * connected by great circle arcs.
 * 
 * Used for geodetic polygons. Points must be in counter-clockwise order when viewed
 * from above the surface. The last point must match the first point to close the ring.
 * 
 * This class handles:
 * - Ray-casting algorithm for point-in-ring tests
 * - Pole containment (rings can contain north/south poles)
 * - Multiple external points for antipodal safety
 * - Antimeridian crossing
 * 
 * Based on cmr.spatial.geodetic-ring Clojure namespace.
 */
public class GeodeticRing {
    
    private static final double EXTERNAL_POINT_PRECISION = 4.0;
    private static final double INTERSECTION_POINT_PRECISION = 5.0;  // Matches Clojure
    private static final double APPROXIMATION_DELTA = Arc.APPROXIMATION_DELTA;
    private static final double MBR_COVERS_TOLERANCE = 0.0000000001;  // Matches Clojure mbr/COVERS_TOLERANCE
    
    private final List<Point> points;
    private final List<Arc> arcs;
    private final Mbr mbr;
    private final List<Point> externalPoints;
    private final boolean containsNorthPole;
    private final boolean containsSouthPole;
    
    /**
     * Private constructor. Use createRing() factory method instead.
     */
    private GeodeticRing(List<Point> points, List<Arc> arcs, Mbr mbr,
                          List<Point> externalPoints, boolean containsNorthPole, 
                          boolean containsSouthPole) {
        this.points = points;
        this.arcs = arcs;
        this.mbr = mbr;
        this.externalPoints = externalPoints;
        this.containsNorthPole = containsNorthPole;
        this.containsSouthPole = containsSouthPole;
    }
    
    /**
     * Creates a geodetic ring from a list of points.
     * Calculates derived fields: arcs, MBR, external points, pole containment.
     * 
     * @param points List of points forming the ring (last must equal first)
     *               Can be either cmr.spatial.shape.Point or cmr.spatial.point.Point (Clojure records)
     * @return A new GeodeticRing
     */
    public static GeodeticRing createRing(List<?> points) {
        // Convert Clojure Points to Java Points if necessary
        List<Point> javaPoints = convertToJavaPoints(points);
        List<Arc> arcs = pointsToArcs(javaPoints);
        
        // Calculate course rotation direction and pole containment
        RotationDirection courseRotation = calculateRotationDirection(arcs);
        boolean containsNorthPole = determineNorthPoleContainment(javaPoints, arcs, courseRotation);
        boolean containsSouthPole = determineSouthPoleContainment(javaPoints, arcs, courseRotation);
        
        // Calculate MBR
        Mbr mbr = calculateMbr(arcs, containsNorthPole, containsSouthPole, javaPoints);
        
        // Calculate external points
        List<Point> externalPoints = calculateExternalPoints(mbr, containsNorthPole, containsSouthPole);
        
        return new GeodeticRing(javaPoints, arcs, mbr, externalPoints, 
                                 containsNorthPole, containsSouthPole);
    }
    
    /**
     * Converts a list of objects (potentially Clojure Point records) to Java Points.
     */
    private static List<Point> convertToJavaPoints(List<?> points) {
        List<Point> result = new ArrayList<>();
        for (Object p : points) {
            if (p instanceof Point) {
                result.add((Point) p);
            } else {
                // Assume it's a Clojure Point record with lon/lat fields
                try {
                    java.lang.reflect.Field lonField = p.getClass().getField("lon");
                    java.lang.reflect.Field latField = p.getClass().getField("lat");
                    double lon = ((Number) lonField.get(p)).doubleValue();
                    double lat = ((Number) latField.get(p)).doubleValue();
                    result.add(new Point(lon, lat));
                } catch (Exception e) {
                    throw new IllegalArgumentException(
                        "Could not convert point to Java Point: " + p.getClass().getName(), e);
                }
            }
        }
        return result;
    }
    
    /**
     * Converts a list of points to a list of arcs.
     */
    private static List<Arc> pointsToArcs(List<Point> points) {
        List<Arc> arcs = new ArrayList<>();
        for (int i = 0; i < points.size() - 1; i++) {
            Point p1 = points.get(i);
            Point p2 = points.get(i + 1);
            
            // Skip creating arc if points are equal (would throw exception)
            if (!pointsEqual(p1, p2)) {
                try {
                    arcs.add(Arc.createArc(p1, p2));
                } catch (IllegalArgumentException e) {
                    // Skip antipodal points or other invalid arcs
                }
            }
        }
        return arcs;
    }
    
    /**
     * Calculates the rotation direction of the ring.
     */
    private static RotationDirection calculateRotationDirection(List<Arc> arcs) {
        if (arcs.isEmpty()) {
            return RotationDirection.NONE;
        }
        
        // Collect all initial and ending courses from each arc
        List<Double> courses = new ArrayList<>();
        for (Arc arc : arcs) {
            courses.add(arc.getInitialCourse());
            courses.add(arc.getEndingCourse());
        }
        
        // Add the first course again to complete the turn
        courses.add(courses.get(0));
        
        // Calculate net bearing change by summing angle deltas
        double netTurn = 0.0;
        for (int i = 0; i < courses.size() - 1; i++) {
            netTurn += angleDelta(courses.get(i), courses.get(i + 1));
        }
        
        // Counter-clockwise: ~360 degrees, Clockwise: ~-360 degrees, None: ~0 degrees
        if (Math.abs(netTurn) < 0.01) {
            return RotationDirection.NONE;
        } else if (netTurn > 0.0) {
            return RotationDirection.COUNTER_CLOCKWISE;
        } else {
            return RotationDirection.CLOCKWISE;
        }
    }
    
    /**
     * Find the difference between a pair of angles (in degrees).
     * Matches the Clojure angle-delta function.
     */
    private static double angleDelta(double a1, double a2) {
        double a2Shifted = a2;
        
        // Shift angle 2 so it is always greater than angle 1
        if (a2 < a1) {
            a2Shifted = a2 + 360.0;
        }
        
        // Calculate left turn amount
        double leftTurn = a2Shifted - a1;
        
        // Determine which is smaller: turning left or turning right
        if (leftTurn == 180.0) {
            // Can't determine which is smaller, return 0
            return 0.0;
        } else if (leftTurn > 180.0) {
            // Turning right is smaller (returns negative value)
            return leftTurn - 360.0;
        } else {
            return leftTurn;
        }
    }
    
    /**
     * Determines if the ring contains the north pole.
     */
    private static boolean determineNorthPoleContainment(List<Point> points, List<Arc> arcs, 
                                                         RotationDirection rotation) {
        // Check if any point is the north pole
        for (Point p : points) {
            if (isNorthPole(p)) {
                return true;
            }
        }
        
        // Check if any arc crosses the north pole
        for (Arc arc : arcs) {
            if (arc.crossesNorthPole()) {
                return true;
            }
        }
        
        // If rotation is NONE, check longitude rotation direction
        if (rotation == RotationDirection.NONE) {
            RotationDirection lonRotation = calculateLongitudeRotation(points);
            return lonRotation == RotationDirection.COUNTER_CLOCKWISE;
        }
        
        return false;
    }
    
    /**
     * Determines if the ring contains the south pole.
     */
    private static boolean determineSouthPoleContainment(List<Point> points, List<Arc> arcs,
                                                         RotationDirection rotation) {
        // Check if any point is the south pole
        for (Point p : points) {
            if (isSouthPole(p)) {
                return true;
            }
        }
        
        // Check if any arc crosses the south pole
        for (Arc arc : arcs) {
            if (arc.crossesSouthPole()) {
                return true;
            }
        }
        
        // If rotation is NONE, check longitude rotation direction
        if (rotation == RotationDirection.NONE) {
            RotationDirection lonRotation = calculateLongitudeRotation(points);
            return lonRotation == RotationDirection.CLOCKWISE;
        }
        
        return false;
    }
    
    /**
     * Calculates the rotation direction based on longitudes only.
     */
    private static RotationDirection calculateLongitudeRotation(List<Point> points) {
        if (points.size() < 2) {
            return RotationDirection.NONE;
        }
        
        // Extract longitudes
        List<Double> longitudes = new ArrayList<>();
        for (Point p : points) {
            longitudes.add(p.getLon());
        }
        
        // Calculate net rotation using angle deltas
        double netTurn = 0.0;
        for (int i = 0; i < longitudes.size() - 1; i++) {
            netTurn += angleDelta(longitudes.get(i), longitudes.get(i + 1));
        }
        
        // Determine direction
        if (Math.abs(netTurn) < 0.01) {
            return RotationDirection.NONE;
        } else if (netTurn > 0.0) {
            return RotationDirection.COUNTER_CLOCKWISE;
        } else {
            return RotationDirection.CLOCKWISE;
        }
    }
    
    /**
     * Calculates the minimum bounding rectangle for the ring.
     */
    private static Mbr calculateMbr(List<Arc> arcs, boolean containsNorthPole,
                                   boolean containsSouthPole, List<Point> points) {
        if (arcs.isEmpty()) {
            return new Mbr(-180.0, 90.0, 180.0, -90.0);
        }
        
        Mbr result = null;
        
        // Union all arc MBRs
        for (Arc arc : arcs) {
            Mbr[] arcMbrs = arc.getMbrs();
            for (Mbr arcMbr : arcMbrs) {
                if (result == null) {
                    result = arcMbr;
                } else {
                    result = unionMbrs(result, arcMbr);
                }
            }
        }
        
        // Expand to include north pole if contained but not explicitly crossed
        if (containsNorthPole && !anyPointIsNorthPole(points) && !anyArcCrossesNorthPole(arcs)) {
            result = new Mbr(-180.0, 90.0, 180.0, result.getSouth());
        }
        
        // Expand to include south pole if contained but not explicitly crossed
        if (containsSouthPole && !anyPointIsSouthPole(points) && !anyArcCrossesSouthPole(arcs)) {
            result = new Mbr(-180.0, result.getNorth(), 180.0, -90.0);
        }
        
        return result;
    }
    
    /**
     * Unions two MBRs handling geodetic wrapping.
     * Faithful translation of cmr.spatial.mbr/union.
     */
    private static Mbr unionMbrs(Mbr m1, Mbr m2) {
        double north = Math.max(m1.getNorth(), m2.getNorth());
        double south = Math.min(m1.getSouth(), m2.getSouth());
        
        boolean m1Crosses = m1.getWest() > m1.getEast();
        boolean m2Crosses = m2.getWest() > m2.getEast();
        
        double west, east;
        
        if (m1Crosses && m2Crosses) {
            // Both cross antimeridian
            west = Math.min(m1.getWest(), m2.getWest());
            east = Math.max(m1.getEast(), m2.getEast());
            
            // If result would cover whole world, set it to that
            if (west <= east) {
                west = -180.0;
                east = 180.0;
            }
        } else if (m1Crosses || m2Crosses) {
            // One crosses the antimeridian
            // Make m1 be the one that crosses
            Mbr crossing = m1Crosses ? m1 : m2;
            Mbr notCrossing = m1Crosses ? m2 : m1;
            
            double w1 = crossing.getWest();
            double e1 = crossing.getEast();
            double w2 = notCrossing.getWest();
            double e2 = notCrossing.getEast();
            
            // We could expand m1 to the east or to the west. Pick the shorter of the two.
            double westDist = w1 - w2;
            double eastDist = e2 - e1;
            
            if (westDist <= 0.0 || eastDist <= 0.0) {
                // Non-crossing is already contained
                west = w1;
                east = e1;
            } else if (eastDist < westDist) {
                // Expand east
                west = w1;
                east = e2;
            } else {
                // Expand west
                west = w2;
                east = e1;
            }
            
            // If result would cover whole world, set it to that
            if (west <= east) {
                west = -180.0;
                east = 180.0;
            }
        } else {
            // Neither crosses the antimeridian
            // Ensure m1.west <= m2.west
            if (m1.getWest() > m2.getWest()) {
                Mbr temp = m1;
                m1 = m2;
                m2 = temp;
            }
            
            double w1 = m1.getWest();
            double e1 = m1.getEast();
            double w2 = m2.getWest();
            double e2 = m2.getEast();
            
            west = Math.min(w1, w2);
            east = Math.max(e1, e2);
            
            // Check if it's shorter to cross the antimeridian
            double dist = east - west;
            double altWest = w2;
            double altEast = e1;
            double altDist = (180.0 - altWest) + (altEast - (-180.0));
            
            if (altDist < dist) {
                west = altWest;
                east = altEast;
            }
        }
        
        return new Mbr(west, north, east, south);
    }
    
    /**
     * Calculates external points (points guaranteed to be outside the ring).
     * Used for ray-casting algorithm.
     * Matches the logic in cmr.spatial.mbr/external-points.
     */
    private static List<Point> calculateExternalPoints(Mbr mbr, boolean containsNorthPole,
                                                       boolean containsSouthPole) {
        List<Point> result = new ArrayList<>();
        
        // Cannot determine external points if ring contains both poles
        if (containsNorthPole && containsSouthPole) {
            return result;
        }
        
        // Find the biggest area around the MBR to place external points
        boolean crossesAntimeridian = mbr.getWest() > mbr.getEast();
        
        double northDist = 90.0 - mbr.getNorth();
        double southDist = mbr.getSouth() - (-90.0);
        double westDist = crossesAntimeridian ? 0.0 : mbr.getWest() - (-180.0);
        double eastDist = crossesAntimeridian ? 0.0 : 180.0 - mbr.getEast();
        
        double biggestDist = Math.max(Math.max(northDist, southDist), 
                                      Math.max(westDist, eastDist));
        
        double w, n, e, s;
        
        if (biggestDist == northDist) {
            // Place points in area north of MBR
            w = -180.0;
            n = 90.0;
            e = 180.0;
            s = mbr.getNorth();
        } else if (biggestDist == southDist) {
            // Place points in area south of MBR
            w = -180.0;
            n = mbr.getSouth();
            e = 180.0;
            s = -90.0;
        } else if (!crossesAntimeridian && biggestDist == westDist) {
            // Place points in area west of MBR
            w = -180.0;
            n = 90.0;
            e = mbr.getWest();
            s = -90.0;
        } else if (!crossesAntimeridian && biggestDist == eastDist) {
            // Place points in area east of MBR
            w = mbr.getEast();
            n = 90.0;
            e = 180.0;
            s = -90.0;
        } else if (crossesAntimeridian) {
            // Place points in area between east and west (the gap when crossing antimeridian)
            w = mbr.getEast();
            n = 90.0;
            e = mbr.getWest();
            s = -90.0;
        } else {
            throw new RuntimeException("Logic error in calculateExternalPoints: " +
                "Could not determine biggest distance area");
        }
        
        // Find 3 points within the area: left, middle, right
        // These are distributed along the longitude range at the mid-latitude
        double midLon = mid(w, e);
        double rightLon = mid(w, midLon);
        double leftLon = mid(midLon, e);
        double midLat = mid(n, s);
        
        result.add(roundPointForExternalPoints(new Point(leftLon, midLat)));
        result.add(roundPointForExternalPoints(new Point(midLon, midLat)));
        result.add(roundPointForExternalPoints(new Point(rightLon, midLat)));
        
        return result;
    }
    
    /**
     * Calculates the midpoint between two values.
     */
    private static double mid(double a, double b) {
        return (a + b) / 2.0;
    }
    
    /**
     * Tests if a point is covered by (inside) this ring using ray-casting algorithm.
     * Matches the Clojure implementation in geodetic_ring.clj exactly.
     * 
     * @param point The point to test
     * @return true if the ring covers the point
     */
    public boolean coversPoint(Point point) {
        // Check if ring contains north or south pole (matches Clojure)
        if (containsNorthPole && isNorthPole(point)) {
            return true;
        }
        if (containsSouthPole && isSouthPole(point)) {
            return true;
        }
        
        // Only do real intersection if the MBR covers the point (matches Clojure)
        if (!geodeticMbrCoversPoint(mbr, point)) {
            return false;
        }
        
        // Check if point is in the point set (matches Clojure)
        for (Point p : points) {
            if (pointsEqual(p, point)) {
                return true;
            }
        }
        
        // Choose an external point for ray-casting
        Point externalPoint = chooseExternalPoint(point);
        if (externalPoint == null) {
            throw new RuntimeException("Could not find suitable external point for ring coverage test. " +
                                     "This can happen when the ring contains both poles.");
        }
        
        // Create the test arc from point to external point (matches Clojure)
        Arc crossingArc;
        try {
            crossingArc = Arc.createArc(point, externalPoint);
        } catch (IllegalArgumentException e) {
            // Points are equal or antipodal - shouldn't happen with proper external point
            return false;
        }
        
        // Find all intersections with ring arcs and round them
        Set<Point> intersections = new HashSet<>();
        for (Arc arc : arcs) {
            List<Point> arcIntersections = ArcLineSegmentIntersections.arcArcIntersections(crossingArc, arc);
            for (Point p : arcIntersections) {
                // Round to 5 decimal places to match Clojure INTERSECTION_POINT_PRECISION
                Point rounded = roundPoint(p);
                intersections.add(rounded);
            }
        }
        
        // Check if point itself is one of the intersections (matches Clojure)
        if (intersections.contains(roundPoint(point))) {
            return true;
        }
        
        // Odd number of intersections = inside, even = outside
        return intersections.size() % 2 == 1;
    }
    
    /**
     * Checks if an MBR covers a point in geodetic space.
     * Matches cmr.spatial.mbr/geodetic-covers-point? and geodetic-lon-range-covers-lon?
     */
    private static boolean geodeticMbrCoversPoint(Mbr mbr, Point point) {
        double lon = point.getLon();
        double lat = point.getLat();
        
        // Check latitude bounds
        if (lat < mbr.getSouth() || lat > mbr.getNorth()) {
            return false;
        }
        
        // Check longitude - matches geodetic-lon-range-covers-lon? exactly
        double west = mbr.getWest() - MBR_COVERS_TOLERANCE;
        double east = mbr.getEast() + MBR_COVERS_TOLERANCE;
        boolean crossesAntimeridian = west > east;
        
        if (crossesAntimeridian) {
            // Crosses antimeridian: longitude is in [west, 180] or [-180, east]
            return lon >= west || lon <= east;
        } else if (Math.abs(lon) == 180.0) {
            // Special case: point is on antimeridian (±180)
            // Check if west or east is within tolerance of ±180
            double within180 = 180.0 - MBR_COVERS_TOLERANCE;
            return Math.abs(west) >= within180 || Math.abs(east) >= within180;
        } else {
            // Normal case: doesn't cross antimeridian
            return lon >= west && lon <= east;
        }
    }
    
    /**
     * Checks if a point is at the North Pole.
     */
    private static boolean isNorthPole(Point p) {
        return Math.abs(p.getLat() - 90.0) < APPROXIMATION_DELTA;
    }
    
    /**
     * Checks if a point is at the South Pole.
     */
    private static boolean isSouthPole(Point p) {
        return Math.abs(p.getLat() + 90.0) < APPROXIMATION_DELTA;
    }
    
    /**
     * Chooses an external point that is not equal or antipodal to the test point.
     * Matches choose-external-point in Clojure.
     */
    private Point chooseExternalPoint(Point point) {
        // Round to EXTERNAL_POINT_PRECISION (4) to match Clojure
        Point rounded = roundPointForExternalPoints(point);
        Point antipodal = antipodalPoint(rounded);
        
        for (Point ep : externalPoints) {
            // Use exact equality since both are rounded to same precision
            if (!pointsExactlyEqual(ep, rounded) && !pointsExactlyEqual(ep, antipodal)) {
                return ep;
            }
        }
        
        return null;
    }
    
    /**
     * Checks if two points are exactly equal (no approximation).
     * Used when comparing rounded points.
     */
    private static boolean pointsExactlyEqual(Point p1, Point p2) {
        return p1.getLon() == p2.getLon() && p1.getLat() == p2.getLat();
    }
    
    /**
     * Returns the antipodal point (opposite side of the sphere).
     */
    private static Point antipodalPoint(Point p) {
        double lon = p.getLon();
        double lat = p.getLat();
        
        double antiLon = lon + 180.0;
        if (antiLon > 180.0) antiLon -= 360.0;
        
        return new Point(antiLon, -lat);
    }
    
    /**
     * Rounds a point for external point generation (precision 4).
     * Matches EXTERNAL_POINT_PRECISION in Clojure.
     */
    private static Point roundPointForExternalPoints(Point p) {
        double scale = Math.pow(10, EXTERNAL_POINT_PRECISION);
        double lon = Math.round(p.getLon() * scale) / scale;
        double lat = Math.round(p.getLat() * scale) / scale;
        return new Point(lon, lat);
    }
    
    /**
     * Rounds a point for intersection calculations (precision 5).
     * Matches INTERSECTION_POINT_PRECISION in Clojure.
     */
    private static Point roundPoint(Point p) {
        double scale = Math.pow(10, INTERSECTION_POINT_PRECISION);  // 100000.0
        double lon = Math.round(p.getLon() * scale) / scale;
        double lat = Math.round(p.getLat() * scale) / scale;
        return new Point(lon, lat);
    }
    
    /**
     * Returns true if any point in the list is at the north pole.
     */
    private static boolean anyPointIsNorthPole(List<Point> points) {
        for (Point p : points) {
            if (isNorthPole(p)) return true;
        }
        return false;
    }
    
    /**
     * Returns true if any point in the list is at the south pole.
     */
    private static boolean anyPointIsSouthPole(List<Point> points) {
        for (Point p : points) {
            if (isSouthPole(p)) return true;
        }
        return false;
    }
    
    /**
     * Returns true if any arc crosses the north pole.
     */
    private static boolean anyArcCrossesNorthPole(List<Arc> arcs) {
        for (Arc arc : arcs) {
            if (arc.crossesNorthPole()) return true;
        }
        return false;
    }
    
    /**
     * Returns true if any arc crosses the south pole.
     */
    private static boolean anyArcCrossesSouthPole(List<Arc> arcs) {
        for (Arc arc : arcs) {
            if (arc.crossesSouthPole()) return true;
        }
        return false;
    }
    
    /**
     * Returns true if two points are equal (with tolerance).
     */
    private static boolean pointsEqual(Point p1, Point p2) {
        return MathUtils.doubleApprox(p1.getLon(), p2.getLon(), APPROXIMATION_DELTA) &&
               MathUtils.doubleApprox(p1.getLat(), p2.getLat(), APPROXIMATION_DELTA);
    }
    
    // Getters
    
    public List<Point> getPointsList() {
        return points;
    }
    
    public Point[] getPoints() {
        return points.toArray(new Point[0]);
    }
    
    public List<Arc> getArcsList() {
        return arcs;
    }
    
    public Arc[] getArcs() {
        return arcs.toArray(new Arc[0]);
    }
    
    public Mbr getMbr() {
        return mbr;
    }
    
    public boolean containsNorthPole() {
        return containsNorthPole;
    }
    
    public boolean containsSouthPole() {
        return containsSouthPole;
    }
    
    /**
     * Enum for rotation direction.
     */
    public enum RotationDirection {
        CLOCKWISE,
        COUNTER_CLOCKWISE,
        NONE
    }
    
    @Override
    public String toString() {
        return String.format("GeodeticRing{points=%d, arcs=%d, northPole=%b, southPole=%b}",
                points.size(), arcs.size(), containsNorthPole, containsSouthPole);
    }
}
